src/main/scala/cognite/spark/v1/FileContentRelation.scala

```diff
@@ -0,0 +1,141 @@
+      backend <- Resource.make(
+        IO(
+          AsyncHttpClientFs2Backend
+            .usingClient[IO](SttpClientBackendFactory.create("file content download"), dispatcher))
+      )(backend => backend.close())
+    } yield backend
+
+  private lazy val dataFrame: DataFrame = createDataFrame(sparkSession = sqlContext.sparkSession)
+
+  override def schema: StructType =
+    dataFrame.schema
+
+  def createDataFrame(sparkSession: SparkSession): DataFrame = {
+    val rdd: RDD[String] = new RDD[String](sparkSession.sparkContext, Nil) with Serializable {
+
+      import cognite.spark.v1.CdpConnector.ioRuntime
+
+      val maxParallelism = 1
+      override def compute(split: Partition, context: TaskContext): Iterator[String] = {
+
+        val validUrl = for {
+          mimeType <- client.files.retrieveByExternalId(fileExternalId).map(_.mimeType)
+          downloadLink <- client.files
+            .downloadLink(FileDownloadExternalId(fileExternalId))
+            .map(_.downloadUrl)
+          _ <- IO.pure(
+            if (!downloadLink.startsWith("https")) {
+              throw new CdfSparkException("File storage is not using https protocol")
+            }
+          )
+          isFileWithinLimits <- isFileWithinLimits(downloadLink)
+        } yield {
+          if (mimeType.isDefined && !Seq(
+              "application/jsonlines",
+              "application/x-ndjson",
+              "application/jsonl").contains(mimeType.get))
+            throw new CdfSparkException("Wrong mimetype. Expects application/jsonlines")
+          if (!isFileWithinLimits)
+            throw new CdfSparkException(
+              f"File size above size limit, or file size header absent from head request")
+          downloadLink
+        }
+
+        StreamIterator(
+          Stream
+            .eval(validUrl)
+            .flatMap(readUrlContent),
+          maxParallelism,
+          None
+        )
+
+      }
+
+      override protected def getPartitions: Array[Partition] =
+        Array(CdfPartition(0))
+    }
+    import sparkSession.implicits._
+    val dataset = rdd.toDS()
+
+    if (inferSchema) {
+      sqlContext.sparkSession.read.json(dataset)
+    } else {
+      dataset.toDF()
+    }
+  }
+
+  private def readUrlContent(link: String): Stream[IO, String] =
+    Stream.resource(sttpFileContentStreamingBackendResource).flatMap { backend =>
+      val request = basicRequest
+        .get(uri"$link")
+        .response(asStreamUnsafe(Fs2Streams[IO]))
+
+      Stream.eval(backend.send(request)).flatMap { response =>
+        response.body match {
+          case Right(byteStream) =>
+            byteStream
+              .through(fs2.text.utf8.decode)
+              .through(fs2.text.lines)
+          case Left(error) =>
+            Stream.raiseError[IO](new Exception(s"Error while requesting underlying file: $error"))
+        }
+      }
+    }
+
+  private def isFileWithinLimits(downloadUrl: String): IO[Boolean] = {
```

Comment:
nit: here's opportunity to return size & detailed error, so roughly `Either[Long, FileTooBig | FileSizeUnknown]`

Reply:
done :)